### PR TITLE
[WIP] Refactoring of ClassMethods

### PR DESCRIPTION
**Pull request by zf2timo**
While debugging the ClassMethods class, i thought this can be optimized, so it is easier to understand. 
I focus on splitting big Methods into smaller parts.

### CODE REVIEW

File: src/ClassMethods.php [hunk 1]
```diff
@@ -134,38 +134,7 @@ public function extract($object)
...
-                    $attribute = substr($method, 3);
-                    if (!property_exists($object, $attribute)) {
-                        $attribute = lcfirst($attribute);
-                    }
-                }
-
-                $this->extractionMethodsCache[$objectClass][$method] = $attribute;
-            }
-        }
+        $this->populateExtractionCache($object, $objectClass);
```
Comment (Ocramius) [inline src/ClassMethods.php]: This is an additional method call: please add a check on whether it needs to be called or not

File: src/ClassMethods.php [hunk 2]
```diff
@@ -265,4 +234,103 @@ private function resetCaches()
...
+
+    /**
+     * @param $object
+     * @return Filter\FilterComposite
+     */
+    private function getMethodFilter($object)
+    {
+        $filter = $this->filterComposite;
+        if ($object instanceof Filter\FilterProviderInterface) {
+            $filter = $this->getFilterFrom($object);
```
Comment (Ocramius) [inline src/ClassMethods.php]: Return early

File: src/ClassMethods.php [hunk 3]
```diff
@@ -265,4 +234,103 @@ private function resetCaches()
...
+        }
+    }
+
+    /**
+     * @param $object
+     * @return Filter\FilterComposite
+     */
+    private function getMethodFilter($object)
+    {
+        $filter = $this->filterComposite;
```
Comment (Ocramius) [inline src/ClassMethods.php]: Assignment not needed if returning early

**Comment by Ocramius** [comment]
Was going to review further, but it seems like the change was made for the sake of it. Don't get me wrong: you applied all the possible extraction that you know of. That's nice, but it will be heavy on performance. Before this patch is considered, please write a benchmark for it (in a separate PR).

You can refer to https://github.com/zendframework/zend-servicemanager/tree/master/benchmarks to see how this can be achieved.

**Comment by zf2timo** [comment]
@Ocramius Thank you for your response. I will add the performance Benchmark in the next days.

**Comment by zf2timo** [comment]
A performance Benachmark was already added by @sandrokeil in PR #22

**Comment by sandrokeil** [comment]
Every additional function call has a performance impact. I think this is important here. 
@zf2timo Maybe you should revert one liner functions? /cc @Ocramius

**Comment by sandrokeil** [comment]
We should consider to take over #5 ([zf-fr/hydrator](https://github.com/zf-fr/hydrator))

**Comment by zf2timo** [comment]
I think the PR #22 should be merged, so we can see, if there is an performance impact and how big it is.

**Comment by zf2timo** [comment]
#22 was close due to inactivity. So closing this as well.

**Pull request closed without merging**